### PR TITLE
Fix NPE in anomaly localization

### PR DESCRIPTION
**Pull request by jackiehanyang**
Signed-off-by: jackiehanyang <jkhanjob@gmail.com>

### Description
[Describe what this change achieves]
 
### Issues Resolved
https://github.com/opensearch-project/ml-commons/issues/234
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: ml-algorithms/src/test/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImplTests.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
 import org.opensearch.action.search.SearchResponse;
 import org.opensearch.client.Client;
 import org.opensearch.common.settings.Settings;
+import org.opensearch.index.IndexNotFoundException;
```
Comment (Zhangxunmt) [inline ml-algorithms/src/test/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImplTests.java:24]: I don't see this "IndexNotFoundException" is used in the test. Why add it?

Comment (Zhangxunmt) [inline ml-algorithms/src/test/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImplTests.java:24]: I don't see this "IndexNotFoundException" is used in the test. Why add it?

**Review by Zhangxunmt — APPROVED** [review]

File: ml-algorithms/src/test/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImplTests.java [hunk 2]
```diff
@@ -234,6 +235,15 @@ public void testGetLocalizedResultsForInvalidTimeRange() {
         anomalyLocalizer.getLocalizationResults(input, outputListener);
     }
 
+    @Test(expected = RuntimeException.class)
```
Comment (ylwu-amzn) [inline ml-algorithms/src/test/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImplTests.java:238]: Why expect `RuntimeException`  rather `IndexNotFoundException` ?
You can use `ExpectedException` to assert the exception type and error message, refer to https://github.com/opensearch-project/ml-commons/blob/main/common/src/test/java/org/opensearch/ml/common/dataframe/ColumnValueTest.java#L17

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImpl.java [hunk 3]
```diff
@@ -126,6 +127,10 @@ private void onOverallAggregatesResponse(MultiSearchResponse response, AnomalyLo
         AnomalyLocalizationOutput.Result result = new AnomalyLocalizationOutput.Result();
         List<Map.Entry<Long, Long>> intervals = timeBuckets.getAllIntervals();
         for (int i = 0; i < intervals.size(); i++) {
+            if (response.getResponses()[i].getFailure() instanceof IndexNotFoundException) {
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImpl.java:130]: Suggest check if index exists or not like this https://github.com/opensearch-project/ml-commons/blob/main/plugin/src/main/java/org/opensearch/ml/indices/MLIndicesHandler.java#L90

File: ml-algorithms/src/test/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImplTests.java (hunk 2 above)
Comment (jackiehanyang) [reply ml-algorithms/src/test/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImplTests.java:238]: Yeah was originally checking the specific type of exception in the test, then found out that this test class and the test class you mentioned are all checking `Runtime Exception.class`, so continuing using the same approach. 

Will change it to check specific exception type

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImpl.java [hunk 4]
```diff
@@ -125,18 +132,23 @@ private void onOverallAggregatesResponse(MultiSearchResponse response, AnomalyLo
...
-            AnomalyLocalizationOutput.Bucket bucket = new AnomalyLocalizationOutput.Bucket();
-            bucket.setStartTime(intervals.get(i).getKey());
-            bucket.setEndTime(intervals.get(i).getValue());
-            bucket.setOverallAggValue(value);
-            result.getBuckets().add(bucket);
-        }
 
-        output.getResults().put(agg.getName(), result);
-        getLocalizedEntities(input, agg, result, output, listener);
+        if (clusterService.state().metadata().hasIndex(input.getIndexName())) {
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImpl.java:136]: This seems can't support index pattern like "servicelog*". You can refer to AD code to get concrete indices https://github.com/opensearch-project/anomaly-detection/blob/main/src/main/java/org/opensearch/ad/util/IndexUtils.java#L143
```
        // the original index might be an index expression with wildcards like "log*",
        // so we need to expand the expression to concrete index name
        String[] concreteIndices = indexNameExpressionResolver.concreteIndexNames(state, IndicesOptions.lenientExpandOpen(), indices);
```

File: ml-algorithms/src/test/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImplTests.java [hunk 5]
```diff
@@ -200,8 +233,14 @@ public void setup() {
...
     }
 
+    @AfterClass
+    public static void resetPortCounter() {
+        portGenerator.set(0);
+    }
+
     @Test
     public void testGetLocalizedResultsGivenNoAnomaly() {
+        when(clusterService.state()).thenReturn(testState);
```
Comment (Zhangxunmt) [inline ml-algorithms/src/test/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImplTests.java:243]: Did you just add this line to resolve the "null" problem?

**Review by Zhangxunmt — APPROVED** [review]

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImpl.java [hunk 6]
```diff
@@ -150,6 +153,20 @@ private void onOverallAggregatesResponse(MultiSearchResponse response, AnomalyLo
...
     }
+    
+    protected boolean isIndexExist(String indexName, ActionListener<AnomalyLocalizationOutput> listener) {
+        String[] concreteIndices = indexNameExpressionResolver.concreteIndexNames(clusterService.state(),
+                IndicesOptions.lenientExpandOpen(), indexName);
+        if (concreteIndices == null || concreteIndices.length == 0) {
+            return false;
+        }
+        for(String index : concreteIndices) {
+            if (!clusterService.state().metadata().hasIndex(index)) {
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImpl.java:164]: If "logs*" pattern has 3 concrete index "logs1", "logs2" and "logs3", if "log1" doesn't exist but "log2" and "log3" exist, we should return true as we still have 2 valid index exist. But this code seems will return false unless all three indexes exist.

Comment (jackiehanyang) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImpl.java]: ok, I was confused about this use case. Updated the logic to return true if any valid index exist.

File: ml-algorithms/src/test/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImplTests.java (hunk 5 above)
Comment (jackiehanyang) [reply ml-algorithms/src/test/java/org/opensearch/ml/engine/algorithms/anomalylocalization/AnomalyLocalizerImplTests.java:255]: yes, I was confused before about what should the mocked ClusterService return. I thought it should return a mocked local node, but turns out it's true

**Review by ylwu-amzn — APPROVED** [review]

**Review by Zhangxunmt — APPROVED** [review]

**Pull request merged**